### PR TITLE
Iterate results over pages

### PR DESCRIPTION
**Pull request by jderusse**
Full pagination supported

### CODE REVIEW

File: src/S3/Result/ListObjectsV2Output.php [hunk 1]
```diff
@@ -82,16 +84,31 @@ class ListObjectsV2Output extends Result implements \IteratorAggregate
...
+        }
+        $input = $this->lastRequest;
+        if (!$input instanceof ListObjectsV2Request) {
+            throw new \InvalidArgumentException('missing last request injected in paginated result');
+        }
+        $page = $this;
         while (true) {
-            yield from $this->CommonPrefixes;
+            $page->initialize();
+            yield from $page->CommonPrefixes;
```
Comment (Nyholm) [inline src/S3/Result/ListObjectsV2Output.php]: Im not sure if it is better or worse.. But one could do 
```php
yield from $page->getCommonPrefixes(true);
```

That would remove the need for `$page->initialize();`

File: src/S3/S3Client.php [hunk 2]
```diff
@@ -324,7 +324,7 @@ public function listObjectsV2($input): ListObjectsV2Output
             $this->getEndpoint($input->requestUri(), $input->requestQuery())
         );
 
-        return new ListObjectsV2Output($response);
+        return new ListObjectsV2Output($response, $this->httpClient, $this, $input);
```
Comment (Nyholm) [inline src/S3/S3Client.php]: HttpClient is not really needed, is it?

Comment (Nyholm) [reply src/S3/S3Client.php]: Oh, yes. it is of course. It is used for the Steamable response. 

Sorry for the noise.

**Comment by jderusse** [comment]
SQS's ListQueues has a configuration for pagination, but everything is returned in the first page (there is no "nextMarker" or thing like that)

File: src/S3/Result/ListObjectsV2Output.php (hunk 1 above)
Comment (jderusse) [reply src/S3/Result/ListObjectsV2Output.php]: good catch, I'll fix that.

**Comment by jderusse** [comment]
Don't merge it, I'll had an integration test

**Comment by jderusse** [comment]
RFR

pagination is not fully tested (currentPage vs LazyLoad) because not implemented in FakeS3 client, and we don't have (yet) endpoint that use paginated results...

**Pull request merged**